Sources/Basic/ArgumentParser.swift

```diff
@@ -0,0 +1,264 @@
+public final class PositionalArgument<T>: CommandLineArgument {
+    override init(name: String, kind: ArgumentKind.Type, usage: String?, isArray: Bool) {
+        precondition(CommandLineArgument.isPositional(arg: name))
+        super.init(name: name, kind: kind, usage: usage, isArray: false)
+    }
+}
+
+/// Argument parser struct reponsible to parse the provided array of arguments and return
+/// the parsed result.
+public struct ArgumentParser {
+    /// A struct representing result of the parsed arguments.
+    public struct Result {
+        /// Internal representation of arguments mapped to their values.
+        private var results = [CommandLineArgument: Any]()
+
+        /// Adds a result.
+        ///
+        /// - parameter
+        ///     - argument: The argument for which this result is being added.
+        ///     - value: The associated value of the argument.
+        ///
+        /// - throws: ArgumentParserError
+        fileprivate mutating func addResult(for argument: CommandLineArgument, result: ArgumentKind) throws {
+            if argument.isArray {
+                var array = [ArgumentKind]()
+                // Get the previously added results if present.
+                if let previousResult = results[argument] as? [ArgumentKind] {
+                    array = previousResult
+                }
+                array.append(result)
+                results[argument] = array
+            } else {
+                results[argument] = result
+            }
+        }
+
+        /// Get an option argument's value from the results.
+        ///
+        /// Since the options are optional, their result may or may not
+        /// be present.
+        public func get<T>(_ arg: OptionArgument<T>) -> T? {
+            return results[arg] as? T
+        }
+
+        /// Get a positional argument's value.
+        public func get<T>(_ arg: PositionalArgument<T>) -> T {
+            return results[arg] as! T
+        }
+    }
+
+    /// List of arguments added to this parser.
+    private var options = [CommandLineArgument]()
+    private var positionalArgs = [CommandLineArgument]()
+
+    public init() {
+    }
+
+    /// Adds an option to the parser.
+    public mutating func add<T: ArgumentKind>(option: String, kind: T.Type, usage: String? = nil) -> OptionArgument<T> {
+        let arg = OptionArgument<T>(name: option, kind: kind, usage: usage, isArray: false)
+        options.append(arg)
+        return arg
+    }
+
+    /// Add array argument type.
+    public mutating func add<T: ArgumentKind>(option: String, kind: Array<T>.Type, usage: String? = nil) -> OptionArgument<Array<T>> {
+        let arg = OptionArgument<Array<T>>(name: option, kind: kind.Element.self, usage: usage, isArray: true)
+        options.append(arg)
+        return arg
+    }
+
+    /// Adds an argument to the parser.
+    public mutating func add<T: ArgumentKind>(argument: String, kind: T.Type, usage: String? = nil) -> PositionalArgument<T> {
```

Comment:
I think `add(positional:`) would be clearer than `argument`, although I don't really like either.